File: measure-android/measure/src/main/java/sh/measure/android/exporter/EventExporter.kt
package sh.measure.android.exporter

import androidx.annotation.VisibleForTesting
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.storage.Database
import sh.measure.android.storage.FileStorage
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Abstraction to run common functions for exporting of events and synchronizing different
 * exporters like [PeriodicEventExporter] and [ExceptionExporter].
 */
internal interface EventExporter {
    fun createBatch(): BatchCreationResult?
    fun getExistingBatches(): LinkedHashMap<String, MutableList<String>>
    fun export(batchId: String, eventIds: List<String>)
}

internal class EventExporterImpl(
    private val logger: Logger,
    private val database: Database,
    private val fileStorage: FileStorage,
    private val networkClient: NetworkClient,
    private val batchCreator: BatchCreator,
) : EventExporter {
    private companion object {
        const val MAX_EXISTING_BATCHES_TO_EXPORT = 30
    }

    @VisibleForTesting
    internal val batchIdsInTransit = CopyOnWriteArrayList<String>()

    override fun export(batchId: String, eventIds: List<String>) {
        if (batchIdsInTransit.contains(batchId)) {
            logger.log(LogLevel.Warning, "Batch $batchId is already in transit, skipping export")
            return
        }
        batchIdsInTransit.add(batchId)
        try {
            val events = database.getEventPackets(eventIds)
            if (events.isEmpty()) {
                // shouldn't happen, but just in case it does we'd like to know.
                logger.log(
                    LogLevel.Error,
                    "No events found for batch $batchId, invalid export request",
                )
                return
            }
            val attachments = database.getAttachmentPackets(eventIds)
            val isSuccessful = networkClient.execute(batchId, events, attachments)
            handleBatchProcessingResult(isSuccessful, batchId, events, attachments)
        } finally {
            // always remove the batch from the list of batches in transit
            batchIdsInTransit.remove(batchId)
        }
    }

    override fun createBatch(): BatchCreationResult? {
        return batchCreator.create()
    }

    override fun getExistingBatches(): LinkedHashMap<String, MutableList<String>> {
        return database.getBatches(MAX_EXISTING_BATCHES_TO_EXPORT)
    }

    private fun handleBatchProcessingResult(
        isSuccessful: Boolean,
        batchId: String,
        events: List<EventPacket>,
        attachments: List<AttachmentPacket>,
    ) {
        if (isSuccessful) {
            val eventIds = events.map { it.eventId }
            database.deleteEvents(eventIds)
            fileStorage.deleteEventsIfExist(eventIds, attachments.map { it.id })
            logger.log(
                LogLevel.Debug,
                "Successfully sent batch $batchId",
            )
        } else {
            logger.log(LogLevel.Error, "Failed to send batch $batchId")
        }
    }
}

File: ios/MeasureSDK/Exporter/EventExporter.swift
//
//  EventExporter.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 17/10/24.
//

import Foundation

protocol EventExporter {
    func createBatch(_ sessionId: String?) -> BatchCreationResult?
    func getExistingBatches() -> [BatchEntity]
    func export(batchId: String, eventIds: [String]) -> HttpResponse?
}

final class BaseEventExporter: EventExporter {
    private let logger: Logger
    private let networkClient: NetworkClient
    private let batchCreator: BatchCreator
    private let batchStore: BatchStore
    private let eventStore: EventStore

    private let maxExistingBatchesToExport = 5
    private var batchIdsInTransit = Set<String>()
    private let lock = NSLock()

    init(logger: Logger, networkClient: NetworkClient, batchCreator: BatchCreator, batchStore: BatchStore, eventStore: EventStore) {
        self.logger = logger
        self.networkClient = networkClient
        self.batchCreator = batchCreator
        self.batchStore = batchStore
        self.eventStore = eventStore
    }

    func createBatch(_ sessionId: String?) -> BatchCreationResult? {
        return batchCreator.create(sessionId: sessionId)
    }

    func getExistingBatches() -> [BatchEntity] {
        return batchStore.getBatches(maxExistingBatchesToExport)
    }

    func export(batchId: String, eventIds: [String]) -> HttpResponse? {
        lock.lock()
        defer { lock.unlock() }

        guard !batchIdsInTransit.contains(batchId) else {
            logger.internalLog(level: .warning, message: "Batch \(batchId) is already in transit, skipping export", error: nil, data: nil)
            return nil
        }

        batchIdsInTransit.insert(batchId)
        defer { batchIdsInTransit.remove(batchId) }

        guard let events = eventStore.getEvents(eventIds: eventIds), !events.isEmpty else {
            logger.internalLog(level: .error, message: "No events found for batch \(batchId), invalid export request", error: nil, data: nil)
            return nil
        }

        // TODO: add attachments code here
        let response = networkClient.execute(batchId: batchId, events: events)

        handleBatchProcessingResult(response: response, batchId: batchId, events: events)
        return response
    }

    // TODO: update below function to handle attachments
    private func handleBatchProcessingResult(response: HttpResponse, batchId: String, events: [EventEntity]) {
        switch response {
        case .success:
            deleteEvents(batchId: batchId, events: events)
            logger.internalLog(level: .debug, message: "Successfully sent batch \(batchId)", error: nil, data: nil)
        case .error(_):
            deleteEvents(batchId: batchId, events: events)
            logger.internalLog(level: .error, message: "Client error while sending batch \(batchId), dropping the batch", error: nil, data: nil)
        }
    }

    private func deleteEvents(batchId: String, events: [EventEntity]) {
        let eventIds = events.map { $0.id }
        eventStore.deleteEvents(eventIds: eventIds)
        batchStore.deleteBatch(batchId)
    }
}
